### ISSUE TITLE
Automatic rewrite of the append only file not functioning

### ISSUE THREAD

**Issue by sigmaRetrieval** [microsoftarchive/redis#126]
When the feature is enabled, redis fails to rewrite the append only file. 
The relevant log section: 

[7932] 12 Jun 13:27:44.448 \* Background append only file rewriting started by pi
d 9772
[9772] 12 Jun 13:27:44.468 # Write error writing append only file on disk: Inval
id argument
[9772] 12 Jun 13:27:44.468 # rewriteAppendOnlyFile failed in qfork: Invalid argu
ment
[7932] 12 Jun 13:27:44.548 # fork operation complete
[7932] 12 Jun 13:27:44.638 # Background AOF rewrite terminated with error

Any idea what could cause this problem?

Thanks in advance

PS. I'm using the code as of 12 june 9 GMT

**Comment by jepickett** [comment]
I don't know what is happening here. Capture a procmon log when the problem happens so that I can identify the failure:
1.  Grab a copy of Process Monitor (http://technet.microsoft.com/en-us/sysinternals/bb896645)
2.  Launch ProcMon.exe
3.  In the Process Monitor Filter dialog add “Process Name” “contains” “redis-server” then “include” to the list of filters:
   ![processmonitorfilterdialog](https://cloud.githubusercontent.com/assets/5193368/3277340/c019118e-f36c-11e3-8f1d-9f221c9c2638.png)
4.  Ensure the Filter-->Drop Filtered Events menu item is checked
5.  Ensure the File-->Capture Events menu item is checked
6.  Replicate your problem (messages will show up in the Process Monitor window)
7.  Use the File-->Save… menu item to save the file as a .PML file
8.  Compress the file and send it to v-jonpic@microsoft.com

If you run the debug version of redis-server, procmon will be able to capture the stack with symbols when the error occurs.

**Comment by StevenVanDijk** [comment]
I'm a coworker of sigmaRetrieval. I'll try to get the PML to you on Monday. 

Yesterday we tried to compile Redis using the latest version of Cygwin (x64). All went fine and it seemed to solve our problems with saving on our development machine (Windows 7). We moved it to test (Windows Server 2012) and after some time we got the following errors:

```
[37812] 13 Jun 17:25:29.934 * Starting automatic rewriting of AOF on 100% growth
[37812] 13 Jun 17:25:29.981 * Background append only file rewriting started by pid 48568
[48568] 13 Jun 17:25:29.997 # Write error writing append only file on disk: Invalid argument
[48568] 13 Jun 17:25:29.997 # rewriteAppendOnlyFile failed in qfork: Invalid argument
[37812] 13 Jun 17:25:30.091 # fork operation complete
[37812] 13 Jun 17:25:30.216 * Background AOF rewrite terminated with success
[37812] 13 Jun 17:25:30.216 * Parent diff successfully flushed to the rewritten AOF (1929 bytes)
[37812] 13 Jun 17:25:30.231 # Error trying to rename the existing AOF to old tempfile: Broken pipe
```

I understand the Cygwin port is not your problem, but I think it is interesting nonetheless. Again the problem seems to be with forking the process. Since we saw Redis fail on two different ports and two different OS'es (win7 and win'12), we lost confidence in running Redis on our production system.

**Comment by jepickett** [comment]
There are 15 places in rewriteAppendOnlyFile() where a failure would lead to the "Write error writing append only file on disk" message you are seeing. I have enabled AOF writing and I am not seeing the error. What does your config file look like?

I would like to capture a procmon log of the error when it happens on your system. Do the following:
1.  Grab a copy of Process Monitor (http://technet.microsoft.com/en-us/sysinternals/bb896645)
2.  Launch ProcMon.exe
3.  In the Process Monitor Filter dialog add “Process Name” “contains” “redis-server” then “include” to the list of filters:

![processmonitorfilterdialog](https://cloud.githubusercontent.com/assets/5193368/3295749/898701ee-f5c9-11e3-8bb8-a0ff9a384f6d.png)
1. Ensure the Filter- - >Drop Filtered Events menu item is checked
2. Ensure the File- - >Capture Events menu item is checked 
3. Replicate your problem (messages will show up in the Process Monitor window)
4. Use the File- - >Save… menu item to save the file as a .PML file
5. Compress the file and send it to v-jonpic@microsoft.com

**Comment by StevenVanDijk** [comment]
Hoi,

Ok, so I built a fresh new version of the x64 Debug configuration. When I try to start it, I get the following error:

[7992] 17 Jun 08:56:14.909 # QForkMasterInit: system error caught. error code=0x000005af, message=VirtualAllocEx failed.: unknown error

The error code has something to do with the page file, which is 8GB in my case. So I restart my computer and try again. This time it works. I run the following code:

```
    [Fact]
    public void SpamTest()
    {
        const string XmlMessage = "<xml><a>aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa</a></xml>";
        const string XmlMessage2 = "<xml><b></b></xml>";
        var r = new Random();

        // Act
        while (true)
        {
            BufferManager.Instance.AddEntry(r.Next(0, 1000000000).ToString(), XmlMessage);
            BufferManager.Instance.AddEntry(r.Next(0, 1000000000).ToString(), XmlMessage2);
        }
    }

    public void AddEntry(string key, string value)
    {
        using (var redisClient = this.Get<IRedisClient>())
        {
            redisClient.ConnectTimeout = Settings.Default.RedisConnectionTimeout;

            string listId = Convert.ToString(key);
            redisClient.AddItemToList(listId, value);
            redisClient.ExpireEntryIn(
                listId,
                TimeSpan.FromMinutes(AlertNodeConfigurationRepository.Instance.GetMaximumBufferTimeActualEvents()));
        }
    }
```

Btw, I’m using the 3.9.71 version of the ServiceStack Redis client. After a while, I get an exception from the client lib:

ServiceStack.Redis.RedisException
Unable to Connect: sPort: 0
   at ServiceStack.Redis.RedisNativeClient.CreateConnectionError()   at ServiceStack.Redis.RedisNativeClient.SendExpectLong(Byte[][] cmdWithBinaryArgs)   at CommandProcessorService.Import.BufferManager.AddEntry(String key, String value) in BufferManager.cs: line 60   at CommandProcessorService.Test.Import.BufferManagerTests.SpamTest() in BufferManagerTests.cs: line 34

I tried again, this time keeping an eye on my memory commit:

[cid:image002.jpg@01CF8A15.9063CE40]

The spikes are where a bgsave is occurring. I tried opening more applications to decrease the amount of available memory. Then I got the exception again and redis-cli reports “ERR” on a bgsave command. I then restarted the unit test and redis crashed (it didn’t go down completely, it just took all available memory and hung).

The output from redis:

[5976] 17 Jun 09:15:02.375 # Server started, Redis version 2.8.9
[5976] 17 Jun 09:15:02.375 \* The server is now ready to accept connections on port 6379
arguments seen:
qfork
        176, 5976
[5976] 17 Jun 09:15:40.412 # fork operation complete
[5976] 17 Jun 09:15:40.542 \* Background saving terminated with success
[…many of these omitted for brevity…]
arguments seen:
qfork
        176, 5976
[5976] 17 Jun 10:00:36.686 # fork operation complete
[5976] 17 Jun 10:00:36.940 \* Background saving terminated with success
[5976] 17 Jun 10:03:02.847 # BeginForkOperation: system error caught. error code=0x000005af, message=BeginForkOperation: VirtualProtect failed: unkn
own error

[5976] 17 Jun 10:03:02.848 # Can't save in background: fork: Invalid argument
[5976] 17 Jun 10:04:26.868 \* 10000 changes in 60 seconds. Saving...
[5976] 17 Jun 10:04:26.873 # BeginForkOperation: system error caught. error code=0x000005af, message=BeginForkOperation: VirtualProtect failed: unkn
own error

[5976] 17 Jun 10:04:26.873 # Can't save in background: fork: Invalid argument
[5976] 17 Jun 10:04:26.973 \* 10000 changes in 60 seconds. Saving...
[5976] 17 Jun 10:04:26.976 # BeginForkOperation: system error caught. error code=0x000005af, message=BeginForkOperation: VirtualProtect failed: unkn
own error

[5976] 17 Jun 10:04:26.976 # Can't save in background: fork: Invalid argument
[5976] 17 Jun 10:04:27.077 \* 10000 changes in 60 seconds. Saving...
arguments seen:
qfork
        176, 5976
[5976] 17 Jun 10:04:37.469 # BeginForkOperation: system error caught. error code=0x000000cb, message=Forked Process did not respond in a timely manner.: unknown error

[5976] 17 Jun 10:04:37.469 # Can't save in background: fork: Invalid argument

Btw the error codes are:
ERROR_ENVVAR_NOT_FOUND
203 (0xCB)
The system could not find the environment option that was entered.
ERROR_COMMITMENT_LIMIT
1455 (0x5AF)
The paging file is too small for this operation to complete.

Hope this helps! The PML is attached.

Met vriendelijke groet,

Steven van Dijk
Manager Technology & Innovation

Mobile +31 (0)6 519 985 73
mailto:first.lastname@ortec.comORTEC
ORTEC Consulting Group | Houtsingel 5 | PO Box 75 | 2700 AB Zoetermeer | The Netherlands
Tel. +31 (0)88 678 3265 | info@ortec.commailto:info@ortec.com | www.ortec-consulting.comhttp://www.ortec-consulting.com/

**Comment by jepickett** [comment]
Redis allocates a memory mapped file for sharing with the fork()ed processes used during AOF + RDB file creation. The default size of this file is equal to the size of physical memory on your system. During the forking process the system swap file is used manage Copy on Write pages created in both the master and the slave process. Thus the system swap file must be larger than size of the memory mapped file. Windows by default will create a swap file that will expand up to about 3.5 times the size of physical memory. In the worst case scenario the swap file commitment level required by Redis and its forked subprocess is 3 times the size of the memory mapped file. In the default case on a server only used for Redis, the default swap file size will be just large enough.  

It is possible that there is something else using a significant portion of the swap space on your system. The 0x5AF (ERROR_COMMITMENT_LIMIT) error points in this direction. In more recent versions of the Microsoft C-runtime library, HeapAlloc() appears to use system swap space directly rather than allocating memory using VirtualAlloc(). There is a performance advantage for doing so, but it also means that swap file blowout is a possibility. Your options for fixing this is to either use the --maxheap flag to limit the size of the heap created in Redis, or to increase the system swap space available.

In src\Win32_Interop\Win32_QFork.cpp there are comments describing how the QFork operation works.It is useful to understand this code when trying to configure your system for maximum robustness while running Redis.

I am not seeing the 0x5AF error in the PML file you sent me. In that file it looks like all the fork() operations all completed successfully. The PML file also does not contain the 0xCB error. I have never seen this error before.

**Comment by StevenVanDijk** [comment]
Mmmmm…I do indeed have my pagefile set to a fixed size of twice my amount of RAM. My collegue (who first reported the problem) has it set to its default settings though, so maybe I reported a different issue then the one he reported.

I’ll try to get him to reproduce the issue and send you a report.

Met vriendelijke groet,

Steven van Dijk
Manager Technology & Innovation

Mobile +31 (0)6 519 985 73
mailto:first.lastname@ortec.comORTEC
ORTEC Consulting Group | Houtsingel 5 | PO Box 75 | 2700 AB Zoetermeer | The Netherlands
Tel. +31 (0)88 678 3265 | info@ortec.commailto:info@ortec.com | www.ortec-consulting.comhttp://www.ortec-consulting.com/

**Comment by jepickett** [comment]
In commit 52c98f5cb81e4d904f3d7db6da250fb5e1a61519 I fixed a bug that should correct this issue. Please reopen if you are still seeing problems.

**Comment by sigmaRetrieval** [comment]
Issue still remains in the release x64 version, but appears not to be present in the debug version. I did send a mail to v-jonpic@microsoft.com with the requested dump.

**Comment by jepickett** [comment]
Please try the binaries included with the latest commit (4e71fc1c6fa30fcef4de5274b2bb3939621c1585). I found that there were a few locations where include file ordering prevented our fwrite() implementation from being called. This was failing in the release build becasue of a race between the windows memory manager loading the memory mapped file view and fwrite(). Let me know if this works for you.

**Comment by jepickett** [comment]
If this is still a problem, please reopen.
